mcf-browser/src/back-end/graph.js

```diff
@@ -0,0 +1,318 @@
+  getRemotePropertyLabels,
+  getRemotePropertyValues,
+  getValueFromValueObj,
+} from './utils.js';
+
+/** Class representation of a single Node in the KG. */
+class Node {
+  /**
+   * Universal id of the node. Every Node obj has a localId, which is either a
+   * local reference used in an uplaoded file, or the dcid if the node is never
+   * referred to locally (remote only Node).
+   * @type {string}
+   */
+  localId;
+  /**
+   * Whether triples from the remote Data Commons Knowledge Graph have already
+   * been fetched.
+   * @type {boolean}
+   */
+  alreadyFetched;
+  /**
+   * Whether the node exists in the Data Commons Knowledge Graph.
+   * @type {boolean}
+   */
+  existsInKG;
+  /**
+   * Dcid of the node. Set only if a remote id is referred to in a local file or
+   * pulled from the Data Commons Knowledge Graph.
+   * @type {?string}
+   */
+  dcid;
+  /**
+   * The head of a linked list of Assertion objects representing the outgoing
+   * triples of the Node object.
+   * @type {?Assertion}
+   */
+  assertions;
+  /**
+   * The head of a linked list of Assertion objects representing the incoming
+   * triples of the Node object.
+   * @type {?Assertion}
+   */
+  invAssertions;
+
+  /**
+   * Create a Node based on a given id.
+   * @param {string} id The id of the node to create.
+   */
+  constructor(id) {
+    this.localId = id;
+    this.alreadyFetched = false;
+    this.existsInKG = false;
+    this.dcid = null;
+    this.assertions = null;
+    this.invAssertions = null;
+  }
+
+  /**
+   * Indicates if a given object is an instance of Node class.
+   * @param {Object} obj The object to check.
+   * @return {boolean} True if the object is an instance of Node.
+   */
+  static isNode(obj) { return obj instanceof Node; }
+
+  /**
+   * Returns a node with the given ID, creates a new node if shouldCreate is
+   * true, or null if the node does not exist and a node should not be created.
+   *
+   * @param {string} id The id of the node to find.
+   * @param {boolean} shouldCreate True if a new Node should be created if it
+   *     does not already exist.
+   * @return {Node|null} The found node if it exists or is created.
+   */
+  static getNode(id, shouldCreate) {
+    const existing = Node.nodeHash[id];
+    if (existing) {
+      return existing;
+    }
+
+    if (shouldCreate) {
+      const newNode = new Node(id);
+      Node.nodeHash[id] = newNode;
+      return newNode;
+    }
+    return null;
+  }
+
+  /**
+   * Returns the reference to the node that is displayed in browser. If the
+   * node has a dcid, then the dcid will be displayed. If the node's local id
+   * is different, then the local id is also displayed.
+   * Ex: <dcid> [l:<localId>]
+   * @return {string} The reference to the node to be displayed.
+   */
+  getRef() {
+    const dcidRef = this.dcid ? this.dcid : '';
+    let localRef = '';
+
+    if (!this.dcid || this.dcid !== this.localId) {
+      localRef = '[l:' + this.localId + ']';
```

Comment:
nit: do we need [] when dcid is missing?

Reply:
not necessarily. I had kept it in just to make it very clear when a dcid was missing. For example in a list of 
(id1, id2 [l:localId], l:localId2)  I thought if we kept the brackets around the last id, then it makes it more obvious that the dcid is missing. 
I will remove the brackets when the dcid is missing if you think that would be better.